eras/shelley/impl/src/Cardano/Ledger/Shelley/Transition.hs

```diff
@@ -0,0 +1,454 @@
+  default tcInitialPParamsG ::
+    ( EraTransition (PreviousEra era)
+    , TranslateEra era PParams
+    , TranslationError era PParams ~ Void
+    ) =>
+    SimpleGetter (TransitionConfig era) (PParams era)
+  tcInitialPParamsG =
+    to $ \tc ->
+      translateEra' (tc ^. tcTranslationContextL) (tc ^. tcPreviousEraConfigL . tcInitialPParamsG)
+
+instance Crypto c => EraTransition (ShelleyEra c) where
+  newtype TransitionConfig (ShelleyEra c) = ShelleyTransitionConfig
+    { stcShelleyGenesis :: ShelleyGenesis c
+    }
+    deriving (Eq, Show, Generic)
+
+  mkTransitionConfig =
+    error "Impossible: There is no EraTransition instance for ByronEra"
+
+  tcPreviousEraConfigL = notSupportedInThisEraL
+
+  tcTranslationContextL =
+    tcShelleyGenesisL . lens toFromByronTranslationContext setFBTC
+    where
+      setFBTC shelleyGenesis FromByronTranslationContext {..} =
+        shelleyGenesis
+          { sgGenDelegs = fbtcGenDelegs
+          , sgProtocolParams = fbtcProtocolParams
+          , sgMaxLovelaceSupply = fbtcMaxLovelaceSupply
+          }
+
+  tcShelleyGenesisL = lens stcShelleyGenesis (\tc sg -> tc {stcShelleyGenesis = sg})
+
+  tcInitialPParamsG = to (sgProtocolParams . stcShelleyGenesis)
+
+-- | Constructor for the base Shelley `TransitionConfig`
+mkShelleyTransitionConfig :: ShelleyGenesis c -> TransitionConfig (ShelleyEra c)
+mkShelleyTransitionConfig = ShelleyTransitionConfig
+
+-- | Get the initial funds from the `TransitionConfig`. This value must be non-empty
+-- only during testing and benchmarking, it must never contain anything on a real system.
+--
+-- /Warning/ - Should only be used in testing and benchmarking. Will result in an error
+-- when NetworkId is set to Mainnet
+tcInitialFundsL ::
+  (HasCallStack, EraTransition era) =>
+  Lens' (TransitionConfig era) (LM.ListMap (Addr (EraCrypto era)) Coin)
+tcInitialFundsL =
+  tcShelleyGenesisL
+    . lens
+      (\sg -> protectMainnet "InitialFunds" sg null (sgInitialFunds sg))
+      (\sg initialFunds -> sg {sgInitialFunds = initialFunds})
+
+-- | Get the initial staking from the `TransitionConfig`. This value must be non-empty
+-- only during testing and benchmarking, it must never contain anything on a real system.
+--
+-- /Warning/ - Should only be used in testing and benchmarking. Will result in an error
+-- when NetworkId is set to Mainnet
+tcInitialStakingL ::
+  (HasCallStack, EraTransition era) =>
+  Lens' (TransitionConfig era) (ShelleyGenesisStaking (EraCrypto era))
+tcInitialStakingL =
+  tcShelleyGenesisL
+    . lens
+      (\sg -> protectMainnet "InitialStaking" sg (== mempty) (sgStaking sg))
+      (\sg staking -> sg {sgStaking = staking})
+
+protectMainnet :: HasCallStack => String -> ShelleyGenesis c -> (a -> Bool) -> a -> a
+protectMainnet name sg isMainnetSafe m =
+  if sgNetworkId sg == Mainnet && not (isMainnetSafe m)
+    then error $ "Injection of " ++ name ++ " is not possible on Mainnet"
+    else m
```

Comment:
Hey @amesgen I wanted to run by you this extra precaution check I implemented in this PR. It essentially prevents from funds and staking injection on Mainnet. Do you think it is safe to do this check (see it at use sites in this module) or are there tests out there that actually rely on this being available for ShelleyGenesis being configured as Mainnet?

Reply:
Yes, I think this is safe. We only call the functions[^1] using these lenses when someone sets `TestShelleyHardForkAtEpoch: 0` (and potentially also for additional eras), which is always incorrect for mainnet. Hence, I don't see any possibility how this could break any valid use case.

[^1]: Implementation details: This happens [here](https://github.com/input-output-hk/ouroboros-consensus/blob/f743b8fe9e3e8f24822ba503ecf43eab13bce66f/ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Node.hs#L1033), and the initial ledger state (constructed via `injectInitialExtLedgerState`) is not in Byron if and only if someone configured a HF to happen in epoch 0 via `TriggerHardForkAtEpoch 0` which is what `TestShelleyHardForkAtEpoch: 0` is translated into.

Reply:
Perfect. That's what I thought. Thank you.